### DISCUSSION TITLE
Rendering multiple workspaces with renderWorkspace at the same time

### DISCUSSION THREAD

**Discussion by raybbian** [hyprwm/Hyprland#8805] [Q&A]
Hello, I'm trying to use `renderWorkspace` to render multiple active desktops at the same time, similar to [hyprexpo](https://github.com/hyprwm/hyprland-plugins/tree/main/hyprexpo). However, unlike hyprexpo, I would like to render all workspaces to the screen directly instead of storing the workspace previews in framebuffers, then rendering those.

Currently, I am able to render only the current active workspace to a portion of the screen. Any attempts to render other workspaces to other parts of the screen using `renderWorkspace` are unsuccessful. I was not able to find any other code/plugins that hook into `renderWorkspace` for my reference. Is this what `renderWorkspace` allows me to do? Is there something I'm not understanding with my [code](https://github.com/raybbian/hyprtasking)? The relevant part is here:

```cpp
void CHyprtaskingView::render() {
    const PHLMONITOR pMonitor = getMonitor();
    if (pMonitor == nullptr)
        return;

    g_pHyprRenderer->makeEGLCurrent();

    std::vector<WORKSPACEID> workspaces;
    for (auto &ws : g_pCompositor->m_vWorkspaces) {
        if (ws == nullptr)
            continue;
        if (ws->m_pMonitor->ID != monitorID)
            continue;
        // ignore special workspaces for now
        if (ws->m_iID < 1)
            continue;
        workspaces.push_back(ws->m_iID);
    }
    std::sort(workspaces.begin(), workspaces.end());

    timespec time;
    clock_gettime(CLOCK_MONOTONIC, &time);

    CBox viewBox = {{0, 0}, pMonitor->vecPixelSize};
    g_pHyprOpenGL->renderRect(&viewBox, CHyprColor{0, 0, 0, 1.0});

    const PHLWORKSPACE startWorkspace = pMonitor->activeWorkspace;
    startWorkspace->m_bVisible = false;

    double workspaceX = 0.0;
    double workspaceY = 0.0;
    double scale = 0.5;
    double workspaceW = pMonitor->vecPixelSize.x * scale;
    double workspaceH = pMonitor->vecPixelSize.y * scale;
    for (const WORKSPACEID wsID : workspaces) {
        const PHLWORKSPACE pWorkspace = g_pCompositor->getWorkspaceByID(wsID);
        if (pWorkspace == nullptr)
            continue;
        CBox curBox{workspaceX, workspaceY, workspaceW, workspaceH};

        g_pHyprOpenGL->clear(CHyprColor{0, 0, 0, 1.0});

        pMonitor->activeWorkspace = pWorkspace;
        pWorkspace->m_bVisible = true;

        ((tRenderWorkspace)g_pRenderWorkspaceHook->m_pOriginal)(
            g_pHyprRenderer.get(), pMonitor, pWorkspace, &time, curBox);

        pWorkspace->m_bVisible = false;

        workspaceX += workspaceW / scale;
    }

    pMonitor->activeWorkspace = startWorkspace;
    startWorkspace->m_bVisible = true;
}
```

**Comment by raybbian** [comment]
After some experimenting it seems like having a layer surface in the background active allows all the workspaces to render properly.

**Comment by vaxerski** [comment]
> However, unlike hyprexpo, I would like to render all workspaces to the screen directly instead of storing the workspace previews in framebuffers, then rendering those.

...why? The renderer is quite bad with dealing with rendering workspaces at any resolution other than the screen size.

**Reply by raybbian** [reply]
In what ways? I've been able to render workspaces with modified scale w/o too many issues.

As a sidenote: is it "required" to call g_pHyprRenderer->beginRender(...) for each rendered workspace? Hyprexpo does so but I'm not sure if that was simply to change the framebuffer being rendered to. I'm getting some weird artifacts with my setup when moving windows between workspaces (the window movement appears on all workspaces). Would you happen to know why this is?: 

https://github.com/user-attachments/assets/63c3aa5e-8349-457e-97ad-58a6aa76bff0

**Reply by raybbian** [reply]
nvm, fixed by warping movingToWorkspaceAlpha to 1 after workspace change

**Reply by vaxerski** [reply]
> is it "required" to call g_pHyprRenderer->beginRender(...) for each rendered workspace?

with the renderer scheduling rewrite in -git, yeah
